### DISCUSSION TITLE
Display progress in the MacOS dock icon

### DISCUSSION THREAD

**Discussion by danipen** [AvaloniaUI/Avalonia#17144] [General]
Someone knows if is there any library or a managed way of displaying a progress bar in the dock icon within Avalonia?

![image](https://github.com/user-attachments/assets/e9e0cc73-f6cb-4c64-ac25-9a43167b2b4f)

AFAIK it can be done using native interop, but I'm asking just in case is there any other better way to do that.

**Comment by maxkatz6** [comment]
No, only native interop is an option here.

**Reply by maxkatz6** [reply]
By "native" I mean using macOS APIs directly. It can be done in C# though, like via net8.0-macos target.

**Reply by danipen** [reply]
Can you clarify this with any example or doc?

**Reply by maxkatz6** [reply]
There is a little bit info here https://docs.avaloniaui.net/docs/guides/platforms/platform-specific-code/dotnet
Microsoft does have documentation for Apple APIs via their mappings https://learn.microsoft.com/en-us/dotnet/api/?view=xamarin-mac-sdk-14

**Reply by maxkatz6** [reply]
If you don't want to introduce any new targets, pinvoke or native library is your only other two options.

**Comment by hamster620** [comment]
For solution without using ```netX.0-macos``` target, you can refer to library [CarinaStudio.AppBase.MacOS](https://github.com/carina-studio/AppBase/tree/master/MacOS) and the [usage](https://github.com/carina-studio/AppSuiteBase/blob/927e1055c861c4b30076113013fcb6ffceaa364d/Core/AppSuiteApplication.MacOS.cs#L365) in our application.

**Reply by danipen** [reply]
Thank you so much!

**Reply by danipen** [reply]
@hamster620 thanks!

I see that you implemented mostly all the interop part with macOS APIs what is awesome! Just out of curiosity... What was the motivation to write it from scratch instead of using Xamarin.macOS or so?

Anyway, thanks for sharing. Your project is awesome!

**Reply by hamster620** [reply]
> @hamster620 thanks!
> 
> I see that you implemented mostly all the interop part with macOS APIs what is awesome! Just out of curiosity... What was the motivation to write it from scratch instead of using Xamarin.macOS or so?
> 
> Anyway, thanks for sharing. Your project is awesome!

There are 2 main reasons:
1. To keep target frameworks as simple as possible instead of net8.0-windows, net8.0-macos, net9.0-....
2. Don't want to include bunch of interop libraries just because I need to use some OS APIs. (Not sure about macOS but I known that the published package becomes much larger if I use Windows TFM).
